package "stabilitytest"

public class ExceptionStability
{
	public static test(StabilityTest program)
	{
		Console.writeLine("Checking Exception handling...")
		
		Bool worked = false
		
		try
		{
			testException(program)
		}
		catch (Exception e)
		{
			worked = true
		}
		
		if (!worked)
		{
			program.fail("Failed; uncaught Exception")
		}
	}
	
	static testException(StabilityTest program)
	{
		Bool worked = false
		
		try
		{
			Console.write("Checking DivideByZeroException... ")
			
			try
			{
				Int den = 0
				Int i   = 43 / den
			}
			catch (DivideByZeroException e)
			{
				Console.write("ASDF")
				worked = true
				
				Console.writeLine("OK")
				Console.write("Checking StabilityTestException... ")
				
				throw new StabilityTestException()
			}
			
			program.fail("Failed to catch DivideByZeroException")
		}
		catch (StabilityTestException e)
		{
			if (!worked)
			{
				program.fail("Failed; uncaught DivideByZeroException")
			}
			
			Console.writeLine("OK")
			
			throw new Exception()
		}
		
		if (!worked)
		{
			program.fail("Failed; uncaught StabilityTestException")
		}
	}
}